pauliopt/pauli/synthesis/steiner_gray_synthesis.py

```diff
@@ -0,0 +1,536 @@
+        for p2 in [X, Y, Z]:
+            found_pair = True
+            for l in columns_to_use:
+                p_gdt = pp[l][q1]
+                p_gdt2 = pp[l][q2]
+                a_valid = p_gdt in [I, p1]
+                b_valid = p_gdt2 in [I, p2]
+                if a_valid != b_valid:
+                    found_pair = False
+                    break
+            if found_pair:
+                return p1, p2
+
+    return None
+
+
+def pick_best_pair(pp, G: nx.Graph, columns_to_use, qubits):
+    for q_1, q_2 in G.edges:
+        if q_1 in qubits and q_2 in qubits:
+            pairs = find_compatible_pair(pp, q_1, q_2, columns_to_use)
+            if pairs is not None:
+                return pairs, (q_1, q_2)
+    return None
+
+
+def filter_identity_qubits(pp: PauliPolynomial, qubits, columns_to_use):
+    non_identity_qubits = []
+    for q in qubits:
+        if pp[columns_to_use[0]][q] != I:
+            non_identity_qubits.append(q)
+    return non_identity_qubits
+
+
+def update_pair_qubits(
+    pp: PauliPolynomial, c: Circuit, topology, qubits, columns_to_use
+):
+    non_visited_qubits = [q for q in qubits]
+    non_visited_qubits = filter_identity_qubits(pp, non_visited_qubits, columns_to_use)
+    qubit_pairs = pick_best_pair(
+        pp,
+        topology.to_nx.subgraph(non_visited_qubits),
+        columns_to_use,
+        non_visited_qubits,
+    )
+    (p1, p2), (q_1, q_2) = qubit_pairs
+    non_visited_qubits.remove(q_1)
+    non_visited_qubits.remove(q_2)
+    update_gadget_single_column(pp, c, q_1, p1, columns_to_use)
+    update_gadget_single_column(pp, c, q_2, p2, columns_to_use)
+
+    pp.propagate(CX(q_1, q_2), columns_to_use)
+    c.cx(q_1, q_2)
+
+
+def partition_pauli_polynomial(pp: PauliPolynomial, row: int, columns_to_use: list):
+    col_i = []
+    col_x = []
+    col_y = []
+    col_z = []
+    for col in columns_to_use:
+        if pp.pauli_gadgets[col][row] == X:
+            col_x.append(col)
+        elif pp.pauli_gadgets[col][row] == Y:
+            col_y.append(col)
+        elif pp.pauli_gadgets[col][row] == Z:
+            col_z.append(col)
+        elif pp.pauli_gadgets[col][row] == I:
+            col_i.append(col)
+        else:
+            raise ValueError("Invalid Pauli in Partition")
+    return col_i, col_x, col_y, col_z
+
+
+def max_partition_pauli_polynomial(pp: PauliPolynomial, row: int, columns_to_use: list):
+    cols = [[X], [Y], [Z]]
+    cols_i = []
+    for col in columns_to_use:
+        if pp.pauli_gadgets[col][row] == X:
+            cols[0].append(col)
+        elif pp.pauli_gadgets[col][row] == Y:
+            cols[1].append(col)
+        elif pp.pauli_gadgets[col][row] == Z:
+            cols[2].append(col)
+        elif pp.pauli_gadgets[col][row] == I:
+            cols_i.append(col)
+        else:
+            raise ValueError("Invalid Pauli in Weighted")
+
+    cols.sort(key=lambda x: len(x))
+
+    return cols_i, cols[-1][1:], cols[-1][0], cols[0][1:] + cols[1][1:]
```

Comment:
Maybe cols should be a dictionary instead

Reply:
Same argument as above, the point was to first integrate our algorithms